clippy_lints/src/identity_conversion.rs

```diff
@@ -0,0 +1,106 @@
+use rustc::lint::*;
+use rustc::hir::*;
+use rustc::hir::intravisit::{Visitor, walk_expr, NestedVisitorMap};
+use syntax::ast::NodeId;
+use syntax::codemap::Span;
+use utils::{match_trait_method, match_def_path, same_tys, in_macro, snippet, span_lint_and_then};
+use utils::paths;
+
+/// **What it does:** Checks for always-identitical `Into`/`From` conversions.
+///
+/// **Why is this bad?** Redundant code.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// let s: String = format!("hello").into();
+/// ```
+declare_lint! {
+    pub IDENTITY_CONVERSION,
+    Warn,
+    "using always-identical `Into`/`From` conversions"
+}
+
+pub struct IdentityConversion;
+
+impl LintPass for IdentityConversion {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(IDENTITY_CONVERSION)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for IdentityConversion {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'a, 'tcx>,
+        _: intravisit::FnKind<'tcx>,
+        _: &'tcx FnDecl,
+        body: &'tcx Body,
+        _: Span,
+        _: NodeId
+    ) {
+        let mut vis = IdentityConversionVisitor { cx: cx };
+        walk_expr(&mut vis, &body.value);
+    }
+}
+
+/// Visiting expressions except ones expanded from `?` (that contains `From::from`)
```

Comment:
couldn't you use `check_expr` and add an `in_macro` check?

Using a custom visitor breaks fine grained `#[allow]` attributes